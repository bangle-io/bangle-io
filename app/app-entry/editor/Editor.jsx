import { BangleEditor, useEditorState } from '@bangle.dev/react';
import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useRef, useState } from 'react';
import { getScrollParentElement } from 'utils';
import { useWorkspaceContext } from 'workspace-context';

const LOG = false;
let log = LOG ? console.log.bind(console, 'play/Editor') : () => {};

Editor.propTypes = {
  wsPath: PropTypes.string.isRequired,
  editorId: PropTypes.number.isRequired,
};

export function Editor({ editorId, wsPath, extensionRegistry, setEditor }) {
  const { getNote } = useWorkspaceContext();
  // an object which can is used to provide extensions a store unique to this editor instance
  const [uniqueEditorObj] = useState({});
  // Even though the collab extension will reset the content to its convenience
  // preloading the content will give us the benefit of static height, which comes
  // in handy when loading editor with a given scroll position.
  const [initialValue, setInitialDoc] = useState();
  useEffect(() => {
    let destroyed = false;
    getNote(wsPath).then((doc) => {
      if (!destroyed) {
        setInitialDoc(doc);
      }
    });
    return () => {
      destroyed = true;
    };
  }, [getNote, wsPath]);

  useEffect(() => {
    if (initialValue) {
      const scrollParent = getScrollParentElement(editorId);
      const pos = extensionRegistry.editor.initialScrollPos({
        wsPath,
        editorId,
        scrollParent,
        doc: initialValue,
        uniqueEditorObj: uniqueEditorObj,
      });
      if (typeof pos === 'number') {
        scrollParent.scrollTop = pos;
      }
    }
  }, [editorId, wsPath, extensionRegistry, uniqueEditorObj, initialValue]);

  useEffect(() => {
    return () => {
      extensionRegistry.editor.beforeDestroy({
        wsPath,
        editorId,
        uniqueEditorObj: uniqueEditorObj,
      });
    };
  }, [wsPath, editorId, uniqueEditorObj, extensionRegistry]);

  return initialValue ? (
    <EditorInner
      editorId={editorId}
      wsPath={wsPath}
      setEditor={setEditor}
      extensionRegistry={extensionRegistry}
      initialValue={initialValue}
      uniqueEditorObj={uniqueEditorObj}
    />
  ) : null;
}

function EditorInner({
  editorId,
  wsPath,
  extensionRegistry,
  setEditor,
  initialValue,
  uniqueEditorObj,
}) {
  useEffect(() => {
    log('mounting editor', editorId, wsPath);
    return () => {
      log('unmounting editor', editorId, wsPath);
    };
  }, [wsPath, editorId]);

  const plugins = useCallback(() => {
    return extensionRegistry.getPlugins();
  }, [extensionRegistry]);

  const onEditorReady = useCallback(
    (editor) => {
      setEditor(editorId, editor);
      editor.focusView();
    },
    [setEditor, editorId],
  );

  const renderNodeViews = useCallback(
    (nodeViewRenderArg) => {
      const { node, updateAttrs, children, selected } = nodeViewRenderArg;

      return extensionRegistry.renderReactNodeViews({
        nodeViewRenderArg,
        wsPath,
        editorId,
      });
    },
    [extensionRegistry, wsPath, editorId],
  );
  const editorState = useEditorState({
    plugins: plugins,
    pluginMetadata: {
      wsPath,
      editorId,
    },
    specRegistry: extensionRegistry.specRegistry,
    initialValue: initialValue,
    pmStateOpts: {
      selection: extensionRegistry.editor.initialSelection({
        wsPath,
        editorId,
        doc: initialValue,
        uniqueEditorObj: uniqueEditorObj,
      }),
    },
    editorProps: {},
  });

  return (
    <BangleEditor
      state={editorState}
      onReady={onEditorReady}
      renderNodeViews={renderNodeViews}
      focusOnInit={false}
      className="bangle-editor-inner-container"
    >
      {extensionRegistry.renderExtensionEditorComponents({ wsPath, editorId })}
    </BangleEditor>
  );
}
