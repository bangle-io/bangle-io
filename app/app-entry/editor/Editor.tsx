import {
  BangleEditor,
  useEditorState,
  RenderNodeViewsFunction,
} from '@bangle.dev/react';
import { Node } from '@bangle.dev/pm';
import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useRef, useState } from 'react';
import { getScrollParentElement } from 'utils';
import { useWorkspaceContext } from 'workspace-context';
import { ExtensionRegistry } from 'extension-registry';
import type { BangleEditor as CoreBangleEditor } from '@bangle.dev/core';

const LOG = false;
let log = LOG ? console.log.bind(console, 'play/Editor') : () => {};

Editor.propTypes = {
  wsPath: PropTypes.string.isRequired,
  editorId: PropTypes.number.isRequired,
};

export function Editor({
  editorId,
  wsPath,
  extensionRegistry,
  setEditor,
}: {
  extensionRegistry: ExtensionRegistry;
  editorId: number;
  wsPath: string;
  setEditor: (editorId: number, editor: CoreBangleEditor) => void;
}) {
  const { getNote } = useWorkspaceContext();
  // an object which can is used to provide extensions a store unique to this editor instance
  const [uniqueEditorObj] = useState({});
  // Even though the collab extension will reset the content to its convenience
  // preloading the content will give us the benefit of static height, which comes
  // in handy when loading editor with a given scroll position.
  const [initialValue, setInitialDoc] = useState<Node | undefined>();
  useEffect(() => {
    let destroyed = false;
    getNote(wsPath).then((doc) => {
      if (!destroyed) {
        setInitialDoc(doc);
      }
    });
    return () => {
      destroyed = true;
    };
  }, [getNote, wsPath]);

  useEffect(() => {
    if (initialValue) {
      const scrollParent = getScrollParentElement(editorId);
      const pos = extensionRegistry.editor.initialScrollPos({
        wsPath,
        editorId,
        scrollParent,
        doc: initialValue,
        uniqueEditorObj: uniqueEditorObj,
      });
      if (typeof pos === 'number' && scrollParent) {
        scrollParent.scrollTop = pos;
      }
    }
  }, [editorId, wsPath, extensionRegistry, uniqueEditorObj, initialValue]);

  useEffect(() => {
    return () => {
      extensionRegistry.editor.beforeDestroy({
        wsPath,
        editorId,
        uniqueEditorObj: uniqueEditorObj,
      });
    };
  }, [wsPath, editorId, uniqueEditorObj, extensionRegistry]);

  return initialValue ? (
    <EditorInner
      editorId={editorId}
      wsPath={wsPath}
      setEditor={setEditor}
      extensionRegistry={extensionRegistry}
      initialValue={initialValue}
      uniqueEditorObj={uniqueEditorObj}
    />
  ) : null;
}

function EditorInner({
  editorId,
  wsPath,
  extensionRegistry,
  setEditor,
  initialValue,
  uniqueEditorObj,
}: {
  extensionRegistry: ExtensionRegistry;
  editorId: number;
  wsPath: string;
  setEditor: (editorId: number, editor: CoreBangleEditor) => void;
  initialValue: any;
  uniqueEditorObj: any;
}) {
  useEffect(() => {
    log('mounting editor', editorId, wsPath);
    return () => {
      log('unmounting editor', editorId, wsPath);
    };
  }, [wsPath, editorId]);

  const plugins = useCallback(() => {
    return extensionRegistry.getPlugins();
  }, [extensionRegistry]);

  const onEditorReady = useCallback(
    (editor) => {
      setEditor(editorId, editor);
      editor.focusView();
    },
    [setEditor, editorId],
  );

  const renderNodeViews: RenderNodeViewsFunction = useCallback(
    (nodeViewRenderArg) => {
      return extensionRegistry.renderReactNodeViews({
        nodeViewRenderArg,
        wsPath,
        editorId,
      });
    },
    [extensionRegistry, wsPath, editorId],
  );
  const editorState = useEditorState({
    plugins: plugins,
    pluginMetadata: {
      wsPath,
      editorId,
    },
    specRegistry: extensionRegistry.specRegistry,
    initialValue: initialValue,
    pmStateOpts: {
      selection: extensionRegistry.editor.initialSelection({
        wsPath,
        editorId,
        doc: initialValue,
        uniqueEditorObj: uniqueEditorObj,
      }),
    },
    editorProps: {},
  });

  return (
    <BangleEditor
      state={editorState}
      onReady={onEditorReady}
      renderNodeViews={renderNodeViews}
      focusOnInit={false}
      className="bangle-editor-inner-container"
    >
      {extensionRegistry.renderExtensionEditorComponents({ wsPath, editorId })}
    </BangleEditor>
  );
}
