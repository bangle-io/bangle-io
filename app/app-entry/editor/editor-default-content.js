export const defaultContent = {
  type: 'doc',
  content: [
    {
      type: 'heading',
      attrs: {
        level: 2,
      },
      content: [
        {
          type: 'text',
          text: 'Hi there,',
        },
      ],
    },
  ],
};
