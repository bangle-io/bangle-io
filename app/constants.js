export const SPLIT_SCREEN_MIN_WIDTH = parseInt(
  getComputedStyle(document.documentElement).getPropertyValue(
    '--wide-screen-min-width',
  ),
  10,
);
