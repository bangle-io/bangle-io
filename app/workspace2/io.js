import * as idb from 'idb-keyval';

export class FileOps {
  // file
  static async getFile(wsPath) {
    return idb.get(wsPath);
  }

  static async deleteFile(wsPath) {
    const file = await this.getFile(wsPath);
    if (!file) {
      throw new FileError(`File ${wsPath} does not exists`);
    }
    return idb.del(wsPath);
  }

  static async renameFile(wsPath, newWsPath) {
    const file = await this.getFile(wsPath);

    if (!file) {
      throw new FileError(`File ${wsPath} does not exists`);
    }

    const newFile = await this.getFile(newWsPath);

    if (newFile) {
      throw new FileError(`File ${newWsPath} already exists`);
    }

    await this.deleteFile(wsPath);
    return this.createFile(newWsPath, file);
  }

  static async createFile(wsPath, payload) {
    if (await this.getFile(wsPath)) {
      throw new FileError(`File ${wsPath} already exists`);
    }
    return idb.set(wsPath, payload);
  }

  static async updateFile(wsPath, payload) {
    if (!(await this.getFile(wsPath))) {
      throw new FileError(`File ${wsPath} does not exists`);
    }
    return idb.set(wsPath, payload);
  }

  static async listFiles(wsName) {
    const keys = (await idb.keys()) || [];

    return keys.filter((k) => k.startsWith(wsName + ':'));
  }
}

export class FileError extends Error {
  constructor(message) {
    super('FileError: ' + message);
    this.name = 'FileError';
  }
}
export class FilePermissionError extends Error {
  constructor(message) {
    super('FilePermissionError: ' + message);
    this.name = 'FilePermissionError';
  }
}
