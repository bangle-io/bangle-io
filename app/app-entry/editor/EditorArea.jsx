import { EditorManagerContext } from 'editor-manager-context';
import { ExtensionRegistryContext } from 'extension-registry';
import React, { useEffect, useContext, useState } from 'react';
import { CloseIcon } from 'ui-components/index';
import { cx, sleep, useDestroyRef } from 'utils/index';
import { checkFileExists } from 'workspaces/index';
import { resolvePath } from 'ws-path';
import { Editor } from './Editor';
import { EmptyEditorPage } from './EmptyEditorPage';

/**
 * This exists to save a render cycle
 * when incomingWsPath changes to something else
 * and while `checkFileExists` is doing its thing
 * we let the previous wsPath stay, to avoid an unwanted flash.
 * @param {*} incomingWsPath
 * @returns
 */
function useHandleWsPath(incomingWsPath) {
  const [wsPath, updateWsPath] = useState(undefined);
  const [fileExists, updateFileExists] = useState(undefined);

  const destroyedRef = useDestroyRef();

  useEffect(() => {
    if (incomingWsPath) {
      checkFileExists(incomingWsPath).then((r) => {
        if (!destroyedRef.current) {
          updateFileExists(r);
          updateWsPath(incomingWsPath);
        }
      });
    }
    if (incomingWsPath == null && wsPath) {
      updateFileExists(undefined);
      updateWsPath(undefined);
    }
  }, [incomingWsPath, wsPath, destroyedRef]);

  return { fileExists, wsPath };
}

export function EditorArea({
  id,
  className,
  editorId,
  showTabs,
  wsPath: incomingWsPath,
  onClose,
}) {
  const { fileExists, wsPath } = useHandleWsPath(incomingWsPath);
  const { setEditor } = useContext(EditorManagerContext);
  const extensionRegistry = useContext(ExtensionRegistryContext);
  const [showEmptyEditor, updateShowEmptyEditor] = useState(false);

  // prevents unwarranted flash of empty editor by waiting
  // a certain time before showing the editor.
  useEffect(() => {
    let destroyed = false;
    if (!wsPath) {
      sleep(150).then(() => {
        if (!wsPath && !destroyed) {
          updateShowEmptyEditor(true);
        }
      });
    }
    if (wsPath && showEmptyEditor) {
      updateShowEmptyEditor(false);
    }
    return () => {
      destroyed = true;
    };
  }, [wsPath, showEmptyEditor]);

  return (
    <div id={id} className={cx('bangle-editor-area', className)}>
      {wsPath && showTabs ? <Tab wsPath={wsPath} onClose={onClose} /> : null}
      <div className={cx('bangle-editor-container', showTabs && 'has-tabs')}>
        {fileExists && wsPath && extensionRegistry && (
          <>
            <Editor
              // Key is used to reload the editor when wsPath changes
              key={wsPath}
              editorId={editorId}
              wsPath={wsPath}
              setEditor={setEditor}
              extensionRegistry={extensionRegistry}
            />
          </>
        )}
        {wsPath && fileExists === false && (
          <h3 className="text-xl sm:text-3xl lg:text-3xl leading-none font-bold  mb-8">
            üïµÔ∏è‚Äç‚ôÄÔ∏è‚Äç Note "{resolvePath(wsPath).fileName}" was not found
          </h3>
        )}
        {showEmptyEditor && <EmptyEditorPage />}
      </div>
    </div>
  );
}

function Tab({ wsPath, onClose }) {
  return (
    <div className="editor-tab">
      <span>{resolvePath(wsPath).fileName}</span>
      <button type="button" onClick={onClose} className={`focus:outline-none`}>
        <CloseIcon className="h-4 w-4" />
      </button>
    </div>
  );
}
