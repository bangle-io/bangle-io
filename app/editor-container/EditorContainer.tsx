import React, { useCallback, useEffect, useState } from 'react';

import {
  useNsmSlice,
  useNsmSliceState,
  useSerialOperationContext,
} from '@bangle.io/api';
import { useNsmSliceDispatch } from '@bangle.io/bangle-store-context';
import {
  CORE_OPERATIONS_CLOSE_EDITOR,
  CORE_OPERATIONS_TOGGLE_EDITOR_SPLIT,
  CorePalette,
  PRIMARY_EDITOR_INDEX,
} from '@bangle.io/constants';
import { vars } from '@bangle.io/css-vars';
import { Editor } from '@bangle.io/editor';
import { nsmSliceWorkspace } from '@bangle.io/nsm-slice-workspace';
import type {
  EditorIdType,
  EternalVars,
  WsPath,
} from '@bangle.io/shared-types';
import {
  nsmEditorManagerSlice,
  toggleEditing,
} from '@bangle.io/slice-editor-manager';
import { nsmNotification } from '@bangle.io/slice-notification';
import { nsmUI, nsmUISlice } from '@bangle.io/slice-ui';
import { cx, useDestroyRef } from '@bangle.io/utils';
import { fs } from '@bangle.io/workspace-info';
import { resolvePath } from '@bangle.io/ws-path';

import { Editorbar } from './Editorbar';
import { EditorIssueComp } from './EditorIssueComp';

export function EditorContainer({
  widescreen,
  editorId,
  wsPath: incomingWsPath,
  eternalVars,
}: {
  widescreen: boolean;
  editorId: EditorIdType;
  wsPath: WsPath | undefined;
  eternalVars: EternalVars;
}) {
  const { noteExists, wsPath } = useHandleWsPath(incomingWsPath);
  const { openedWsPaths } = useNsmSliceState(nsmSliceWorkspace);
  const { editingAllowed, focusedEditorId } = useNsmSliceState(
    nsmEditorManagerSlice,
  );

  const { dispatchSerialOperation } = useSerialOperationContext();
  const editorStoreDispatch = useNsmSliceDispatch(nsmEditorManagerSlice);

  const [, uiDispatch] = useNsmSlice(nsmUISlice);

  const isSplitEditorOpen = Boolean(openedWsPaths.secondaryWsPath);

  const onPressSecondaryEditor = useCallback(() => {
    dispatchSerialOperation({
      name: CORE_OPERATIONS_TOGGLE_EDITOR_SPLIT,
    });
  }, [dispatchSerialOperation]);

  const onCloseEditor = useCallback(() => {
    dispatchSerialOperation({
      name: CORE_OPERATIONS_CLOSE_EDITOR,
      value: editorId,
    });
  }, [dispatchSerialOperation, editorId]);

  const openNotesPalette = useCallback(() => {
    uiDispatch(nsmUI.togglePalette(CorePalette.Notes));
  }, [uiDispatch]);

  const { editorIssue, onPressEditorIssue } = useEditorIssue(wsPath);

  const onEnableEditing = useCallback(() => {
    editorStoreDispatch(
      toggleEditing({
        editingAllowed: true,
      }),
    );
  }, [editorStoreDispatch]);

  let children;

  if (noteExists === 'NOT_FOUND') {
    children = (
      <div className="flex flex-col justify-center h-full px-3 align-middle">
        <h3 className="mb-8 text-xl font-bold leading-none sm:text-3xl lg:text-3xl">
          üïµÔ∏è‚Äç‚ôÄÔ∏è‚Äç Note "{wsPath ? resolvePath(wsPath).fileName : ''}" was not found
        </h3>
      </div>
    );
  } else if (noteExists === 'NO_WS_PATH') {
    children = <div>Nothing was found here</div>;
  } else if (wsPath) {
    children = (
      <Editor
        editorId={editorId}
        wsPath={wsPath}
        eternalVars={eternalVars}
        className={`B-editor-container_editor B-editor-container_editor-${editorId}`}
      />
    );
  }

  return (
    <div
      className={cx(
        'B-editor-container_editor-container',
        'B-editor-container_editor-container-' + editorId,
        widescreen && 'overflow-y-scroll',
        'w-full h-full flex flex-col items-center',
      )}
      style={{
        backgroundColor: vars.misc.editorBg,
      }}
    >
      {wsPath && (
        <div className="w-full sticky top-0 z-10">
          {widescreen && (
            <Editorbar
              isActive={focusedEditorId === editorId}
              wsPath={wsPath}
              onClose={onCloseEditor}
              showSplitEditor={editorId === PRIMARY_EDITOR_INDEX}
              onPressSecondaryEditor={onPressSecondaryEditor}
              isSplitEditorOpen={isSplitEditorOpen}
              openNotesPalette={openNotesPalette}
              onEnableEditing={onEnableEditing}
              editingDisabled={!editingAllowed}
            />
          )}
          {editorIssue && (
            <EditorIssueComp
              className={widescreen ? '' : 'pt-3'}
              editorIssue={editorIssue}
              onPress={onPressEditorIssue}
            />
          )}
        </div>
      )}

      <div
        className={cx('w-full smallscreen:min-h-screen')}
        style={{
          maxWidth: `min(${vars.misc.pageMaxWidth}, 100vw)`,
          padding: vars.misc.pagePadding,
        }}
      >
        {children}
      </div>
    </div>
  );
}

function useEditorIssue(wsPath: WsPath | undefined) {
  const [, uiDispatch] = useNsmSlice(nsmUISlice);
  const { editorIssues } = useNsmSliceState(
    nsmNotification.nsmNotificationSlice,
  );
  const editorIssue = wsPath
    ? editorIssues.find((e) => e.wsPath === wsPath)
    : undefined;

  const { dispatchSerialOperation } = useSerialOperationContext();

  const onPressEditorIssue = useCallback(() => {
    if (!editorIssue) {
      return;
    }

    const { serialOperation } = editorIssue;

    if (serialOperation) {
      dispatchSerialOperation({ name: serialOperation });
    } else {
      uiDispatch(
        nsmUI.showGenericErrorModal({
          title: editorIssue.title,
          description: editorIssue.description,
        }),
      );
    }
  }, [uiDispatch, dispatchSerialOperation, editorIssue]);

  return { editorIssue, onPressEditorIssue };
}

/**
 * This exists to save a render cycle
 * when incomingWsPath changes to something else
 * and while `checkFileExists` is doing its thing
 * we let the previous wsPath stay, to avoid an unwanted flash.
 * @param {*} incomingWsPath
 * @returns
 */
function useHandleWsPath(incomingWsPath?: WsPath) {
  const [wsPath, updateWsPath] = useState<WsPath | undefined>(undefined);
  const [noteExists, updateFileExists] = useState<
    'LOADING' | 'FOUND' | 'NOT_FOUND' | 'NO_WS_PATH'
  >(incomingWsPath ? 'LOADING' : 'NO_WS_PATH');

  const destroyedRef = useDestroyRef();

  useEffect(() => {
    if (incomingWsPath) {
      updateFileExists('LOADING');
      fs.fileExists(incomingWsPath).then(
        (r) => {
          if (!destroyedRef.current) {
            if (r) {
              updateFileExists('FOUND');
            } else {
              updateFileExists('NOT_FOUND');
            }
            updateWsPath(incomingWsPath);
          }
        },
        (error) => {
          // silence any errors here as other part of code will handle them
        },
      );
    }
    if (incomingWsPath == null && wsPath) {
      updateFileExists('NO_WS_PATH');
      updateWsPath(undefined);
    }
  }, [incomingWsPath, wsPath, destroyedRef]);

  return { noteExists, wsPath };
}
