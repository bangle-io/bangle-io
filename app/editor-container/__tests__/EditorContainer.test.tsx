/**
 * @jest-environment @bangle.io/jsdom-env
 */
import { act, render, waitFor } from '@testing-library/react';
import React from 'react';

import { SECONDARY_EDITOR_INDEX } from '@bangle.io/constants';
import { Editor } from '@bangle.io/editor';
import { useEditorManagerContext } from '@bangle.io/slice-editor-manager';
import {
  checkFileExists,
  useWorkspaceContext,
} from '@bangle.io/slice-workspace';
import {
  getUseEditorManagerContextReturn,
  getUseWorkspaceContextReturn,
} from '@bangle.io/test-utils';

import { EditorContainer } from '../EditorContainer';

jest.mock('@bangle.io/extension-registry', () => {
  const actual = jest.requireActual('@bangle.io/extension-registry');

  return {
    ...actual,
    useExtensionRegistryContext: jest.fn(),
  };
});

jest.mock('@bangle.io/slice-workspace', () => {
  const actual = jest.requireActual('@bangle.io/slice-workspace');

  return {
    ...actual,
    checkFileExists: jest.fn(() => () => Promise.resolve(true)),
    useWorkspaceContext: jest.fn(),
  };
});

jest.mock('@bangle.io/slice-notification', () => {
  const actual = jest.requireActual('@bangle.io/slice-notification');

  return {
    ...actual,
    getEditorIssue: jest.fn(() => () => undefined),
  };
});

jest.mock('@bangle.io/slice-editor-manager', () => {
  const actual = jest.requireActual('@bangle.io/slice-editor-manager');

  return {
    ...actual,
    useEditorManagerContext: jest.fn(),
  };
});

jest.mock('@bangle.io/editor', () => {
  const actual = jest.requireActual('@bangle.io/editor');

  return {
    ...actual,
    Editor: jest.fn(),
  };
});

jest.mock('../config', () => {
  const actual = jest.requireActual('../config');

  return {
    ...actual,
    EDITOR_LOAD_WAIT_TIME: 0,
  };
});

let useEditorManagerContextMock =
  useEditorManagerContext as jest.MockedFunction<
    typeof useEditorManagerContext
  >;

let checkFileExistsMock = checkFileExists as jest.MockedFunction<
  typeof checkFileExists
>;

let result: ReturnType<typeof render>;

beforeEach(() => {
  (Editor as any).mockImplementation(() => {
    return <div data-testid="mock-editor">MOCK_EDITOR</div>;
  });

  (useWorkspaceContext as any).mockImplementation(() => {
    return { ...getUseWorkspaceContextReturn };
  });

  useEditorManagerContextMock.mockImplementation(() => {
    return {
      ...getUseEditorManagerContextReturn,
    };
  });
  checkFileExistsMock.mockImplementation(() => async () => true);
});

test('basic renders', async () => {
  act(() => {
    result = render(
      <div>
        <EditorContainer
          editorId={SECONDARY_EDITOR_INDEX}
          widescreen={true}
          wsPath="something:blah.md"
        />
      </div>,
    );
  });

  await waitFor(() => {
    expect(result.getByTestId('mock-editor')).toBeTruthy();
  });

  expect(checkFileExists).nthCalledWith(1, 'something:blah.md');

  expect(Editor).lastCalledWith(
    {
      editorId: SECONDARY_EDITOR_INDEX,
      wsPath: 'something:blah.md',
      className: `B-editor-container_editor B-editor-container_editor-1`,
      extensionRegistry: undefined,
    },
    {},
  );
  expect(result!.container).toMatchSnapshot();
});

test('renders correctly when file does not exist', async () => {
  checkFileExistsMock.mockImplementation(() => async () => false);

  act(() => {
    result = render(
      <div>
        <EditorContainer
          editorId={SECONDARY_EDITOR_INDEX}
          widescreen={true}
          wsPath="something:blah.md"
        />
      </div>,
    );
  });

  await result!.findByText('🕵️‍♀️‍ Note "blah.md" was not found');

  expect(result!.container.innerHTML).toContain(`was not found`);
  expect(result!.container).toMatchSnapshot();
});

test('renders when no wsPath is provided', async () => {
  act(() => {
    result = render(
      <div>
        <EditorContainer editorId={1} widescreen={true} wsPath={undefined} />
      </div>,
    );
  });

  await result!.findByText('Nothing was found here');

  expect(checkFileExists).toHaveBeenCalledTimes(0);
  expect(result!.container).toMatchSnapshot();
});

test('changing of wsPath works', async () => {
  act(() => {
    result = render(
      <div>
        <EditorContainer
          editorId={SECONDARY_EDITOR_INDEX}
          widescreen={true}
          wsPath="something:one.md"
        />
      </div>,
    );
  });

  await waitFor(() => {
    expect(result.getByTestId('mock-editor')).toBeTruthy();
  });

  expect(checkFileExists).nthCalledWith(1, 'something:one.md');

  expect(Editor).lastCalledWith(
    {
      editorId: SECONDARY_EDITOR_INDEX,
      wsPath: 'something:one.md',
      className: `B-editor-container_editor B-editor-container_editor-1`,
      extensionRegistry: undefined,
    },
    {},
  );

  act(() => {
    result.rerender(
      <div>
        <EditorContainer
          editorId={SECONDARY_EDITOR_INDEX}
          widescreen={true}
          wsPath="something:two.md"
        />
      </div>,
    );
  });

  await waitFor(() => {
    expect(result.getByTestId('mock-editor')).toBeTruthy();
  });

  expect(Editor).lastCalledWith(
    {
      editorId: SECONDARY_EDITOR_INDEX,
      wsPath: 'something:two.md',
      className: `B-editor-container_editor B-editor-container_editor-1`,
      extensionRegistry: undefined,
    },
    {},
  );
});
