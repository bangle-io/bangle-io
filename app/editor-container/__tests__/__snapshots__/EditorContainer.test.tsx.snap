// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic renders 1`] = `
<div>
  <div>
    <div
      class="B-editor-container_editor-container B-editor-container_editor-container-1 overflow-y-scroll w-full h-full flex flex-col items-center "
    >
      <div
        class="w-full sticky top-0 z-10"
      >
        <div
          class="flex flex-row justify-between w-full px-2 py-1 lg:px-4"
          data-testid="app-editor-container_editorbar"
        >
          <button
            aria-label="note path"
            class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
            style="background-color: transparent;"
            type="button"
          >
            <span
              class="flex flex-grow-1 overflow-hidden "
              style="justify-content: center;"
            >
              <span
                class="text-ellipsis overflow-hidden "
              >
                blah
              </span>
            </span>
          </button>
          <div
            class="flex flex-row flex-1"
          />
          <div
            class="flex flex-row"
          >
            <button
              aria-label="Close"
              class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
              style="background-color: transparent;"
              type="button"
            >
              <span
                class="flex flex-grow-1 overflow-hidden "
                style="justify-content: center;"
              >
                <span>
                  <svg
                    class=" w-4 h-4  "
                    fill="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="w-full smallscreen:min-h-screen "
        style="max-width: min(var(--BV-miscPageMaxWidth), 100vw);"
      >
        <div
          data-testid="mock-editor"
        >
          MOCK_EDITOR
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly when file does not exist 1`] = `
<div>
  <div>
    <div
      class="B-editor-container_editor-container B-editor-container_editor-container-1 overflow-y-scroll w-full h-full flex flex-col items-center "
    >
      <div
        class="w-full sticky top-0 z-10"
      >
        <div
          class="flex flex-row justify-between w-full px-2 py-1 lg:px-4"
          data-testid="app-editor-container_editorbar"
        >
          <button
            aria-label="note path"
            class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
            style="background-color: transparent;"
            type="button"
          >
            <span
              class="flex flex-grow-1 overflow-hidden "
              style="justify-content: center;"
            >
              <span
                class="text-ellipsis overflow-hidden "
              >
                blah
              </span>
            </span>
          </button>
          <div
            class="flex flex-row flex-1"
          />
          <div
            class="flex flex-row"
          >
            <button
              aria-label="Close"
              class="text-xs font-600 min-w-6 h-6 px-1_5  select-none inline-flex justify-center items-center rounded-md whitespace-nowrap overflow-hidden py-1 transition-all duration-100 cursor-pointer "
              style="background-color: transparent;"
              type="button"
            >
              <span
                class="flex flex-grow-1 overflow-hidden "
                style="justify-content: center;"
              >
                <span>
                  <svg
                    class=" w-4 h-4  "
                    fill="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6.2253 4.81108C5.83477 4.42056 5.20161 4.42056 4.81108 4.81108C4.42056 5.20161 4.42056 5.83477 4.81108 6.2253L10.5858 12L4.81114 17.7747C4.42062 18.1652 4.42062 18.7984 4.81114 19.1889C5.20167 19.5794 5.83483 19.5794 6.22535 19.1889L12 13.4142L17.7747 19.1889C18.1652 19.5794 18.7984 19.5794 19.1889 19.1889C19.5794 18.7984 19.5794 18.1652 19.1889 17.7747L13.4142 12L19.189 6.2253C19.5795 5.83477 19.5795 5.20161 19.189 4.81108C18.7985 4.42056 18.1653 4.42056 17.7748 4.81108L12 10.5858L6.2253 4.81108Z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="w-full smallscreen:min-h-screen "
        style="max-width: min(var(--BV-miscPageMaxWidth), 100vw);"
      >
        <div
          class="flex flex-col justify-center h-full px-3 align-middle"
        >
          <h3
            class="mb-8 text-xl font-bold leading-none sm:text-3xl lg:text-3xl"
          >
            üïµÔ∏è‚Äç‚ôÄÔ∏è‚Äç Note "
            blah.md
            " was not found
          </h3>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders when no wsPath is provided 1`] = `
<div>
  <div>
    <div
      class="B-editor-container_editor-container B-editor-container_editor-container-1 overflow-y-scroll w-full h-full flex flex-col items-center "
    >
      <div
        class="w-full smallscreen:min-h-screen "
        style="max-width: min(var(--BV-miscPageMaxWidth), 100vw);"
      >
        <div>
          Nothing was found here
        </div>
      </div>
    </div>
  </div>
</div>
`;
